### ISSUE TITLE
Unnecessary re-renders caused by ReduxForm when rendering using props.children

### ISSUE THREAD

**Issue by stevethatcodes** [erikras/redux-form#3976]
### Are you submitting a **bug report** or a **feature request**?

Bug

### What is the current behavior?

When a redux form decorated component renders anything using `props.children` and the store is updated, a re-render is triggered. Even if the change is in a part of the store unrelated to the form.

### What is the expected behavior?
ReduxForm should only trigger re-renders of decorated components when the form is updated, even when the form is constructed via rendering of `props.children`.

### Sandbox Link

https://codesandbox.io/s/xzj1o7vnjw

### What's your environment?

Node: 9.11
Redux form: 7.3.0
React: 16.3.1
React Redux: 5.0.7
Redux Pack: 0.1.5
Mac OS / Chrome latest

### Other information

* Removing the `{children}` statement in the render method of `form.js` in the sandbox will stop the unnecessary renders.

**Comment by danielrob** [comment]
Please remember to save the codesandbox. The current version isn't pointing to your app.js. 

This is by design after the react 16 update - See #3480.

**Comment by stevethatcodes** [comment]
Apologies, I hadn't used codesandbox before, sorted now, the link in the issue works as expected again.

Is there any recommended approach to mitigate the re-renders? It makes having anything else that's part of `props.children` unable to track its own state outside of the store.

I also wanted to double check that it's expected behaviour for the reduxForm decorated component to update when there are any changes to the store, even to parts not related with the form at all?

**Comment by danielrob** [comment]
Good questions... I think things are slightly more fluid than usual in this area with the introduction of react 16. Could you clarify the impact a little more e.g. 
>  It makes having anything else that's part of props.children unable to track its own state outside of the store.

**Comment by stevethatcodes** [comment]
Sure. The following scenario is where we encountered it. Please excuse the pseudo code but I don't have access to my usual repository today:

```
@reduxForm(...)
<FormComponent>
    <Field name="field_1" />
    <Field name="field_2" />
    <Hierarchy data={[...]} />
</FormComponent>
```

Where `<Hierarchy>` is a component that has expand/collapse states for each node stored in local state.

So the user experience we see, is that when something causes the store to be updated, for example, a global interface setting such as user notification, the state of `<Hierarchy>` is reset and expansion/collapse is lost.

**Comment by shakhbulatgaz** [comment]
@stevethatcodes did you find any workaround or solution for this problem?

**Comment by stevethatcodes** [comment]
> @stevethatcodes did you find any workaround or solution for this problem?

No. We ended up moving to another library in the end so didn't need to solve this specific issue.
